### PR TITLE
feat(upgrade): Add helm hooks to manage CRD upgrades

### PR DESCRIPTION
**Pull request by snehachhabria**
**Description**:

This commit allows OSM to manage the lifecycle of CRD's by adding a pre-install and pre-upgrade helm hook to update the CRD's. The cleanup hook (post-delete helm hook)  has also been modified to ensure that all the CRD's are deleted from the cluster once OSM is uninstalled.

A custom image is built in OSM that contains all the CRD's needed by a particular version and the pod that runs as a pre-install, pre-upgrade helm hook re-applies the CRD's, thus ensuring all the CRD's are upto date. 

Part of #3396

Signed-off-by: Sneha Chhabria <snchh@microsoft.com>

**Affected area**:
| Functional Area            |     |
| -------------------------- | --- |
| New Functionality          | [ ] |
| Documentation              | [ ] |
| Install                    | [X] |
| CLI Tool                   | [ ] |
| Certificate Management     | [ ] |
| Control Plane              | [ ] |
| Ingress                    | [ ] |
| Egress                     | [ ] |
| Networking                 | [ ] |
| Observability              | [ ] |
| SMI Policy                 | [ ] |
| Sidecar Injection          | [ ] |
| Security                   | [ ] |
| Upgrade                    | [X] |
| Tests                      | [ ] |
| CI System                  | [ ] |
| Demo                       | [ ] |
| Performance                | [ ] |
| Other                      | [ ] |

Please answer the following questions with yes/no.

1. Does this change contain code from or inspired by another project? `no`
    -   Did you notify the maintainers and provide attribution?

1. Is this a breaking change? `no`

### CODE REVIEW

**Review by shashankram — APPROVED** [review]

File: charts/osm/templates/cleanup-hook.yaml [hunk 1]
```diff
@@ -9,23 +118,26 @@ metadata:
...
+  containers:
+    - name: garbage-collector
+      image: bitnami/kubectl
+      imagePullPolicy: IfNotPresent
+      command:
+        - sh
+        - -c
+        - >
+         kubectl delete --ignore-not-found meshconfig -n '{{ include "osm.namespace" . }}' osm-mesh-config;
+         kubectl delete --ignore-not-found meshconfig -n '{{ include "osm.namespace" . }}' osm-mesh-config;
```
Comment (allenlsy) [inline charts/osm/templates/cleanup-hook.yaml:135]: Why we delete it twice? Do you mean preset-mesh-config?

Comment (snehachhabria) [reply charts/osm/templates/cleanup-hook.yaml]: it was accidental I removed it.

**Review by eduser25 — APPROVED** [review]

**Comment by ritazh** [comment]
Great PR @snehachhabria! 
> The cleanup hook (post-delete helm hook) has also been modified to ensure that all the CRD's are deleted from the cluster once OSM is uninstalled.

Wouldn't this also remove user-created custom resources?

**Comment by snehachhabria** [comment]
> Great PR @snehachhabria!
> 
> > The cleanup hook (post-delete helm hook) has also been modified to ensure that all the CRD's are deleted from the cluster once OSM is uninstalled.
> 
> Wouldn't this also remove user-created custom resources?

@ritazh that's right, the reason for this is because we have implemented logic to patch a conversion webhook to each of the CRDs (primarily to support older custom resources after OSM has been updated to use the newer CRD). The conversion webhook ties the CRD very tightly to OSM and if we leave the CRD after an uninstall it becomes unusable as k8s will complain about a missing webhook service.

We have an issue to fix this : https://github.com/openservicemesh/osm/issues/3760. I am thinking that on uninstall if the customer explicitly specified not to delete the CRD we will re-patch it to remove the conversion webhook

**Pull request merged**